### PR TITLE
feat: add AbortSignal.timeout(number)

### PR DESCRIPTION
**Pull request by zone117x**
Add `AbortSignal.timeout(number)` definition to `@types/node`, available since Nodejs v17.3.0, v16.14.0.

See https://nodejs.org/api/globals.html#static-method-abortsignaltimeoutdelay for more details.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] [Add or edit tests](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#my-package-teststs) to reflect the change.
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] [Run `npm test <package to test>`](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#running-tests).

Select one of these and delete the others:

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: https://nodejs.org/api/globals.html#static-method-abortsignaltimeoutdelay
- [x] If this PR brings the type definitions up to date with a new version of the JS library, update the version number in the header.

### CODE REVIEW

File: types/node/v14/globals.d.ts [hunk 1]
```diff
@@ -111,11 +111,21 @@ declare var AbortController: {
...
-declare var AbortSignal: {
+interface AbortSignalConstructor {
     prototype: AbortSignal;
     new(): AbortSignal;
+    /**
+     * Returns an AbortSignal that will automatically abort after a specified time.
+     *
+     * @param time The "active" time in milliseconds before the returned AbortSignal will abort.
+     *
+     * @since v17.3.0, v16.14.0
```
Comment (eps1lon) [inline types/node/v14/globals.d.ts:122]: Why add it to v14 when this states it's not available in 14.x?

Comment (zone117x) [reply types/node/v14/globals.d.ts:122]: I'm actually running into issues getting this change to pass the type checker. Any tips?

**Comment by zone117x** [comment]
@peterblazejewicz I'm struggling with getting this to work. It looks like the problem is that `AbortSignal` is defined in `typescript/lib/lib.dom.d.ts` as well as in `@types/node`, so if I add a property I'm getting the error:

```
Subsequent variable declarations must have the same type.  Variable 'AbortSignal' must be of type '{ new (): AbortSignal; prototype: AbortSignal; }', but here has type 'AbortSignalConstructor'.

lib.dom.d.ts(1923, 13): 'AbortSignal' was also declared here.
```

image


Is it possible to make this change without also changing `typescript/lib/lib.dom.d.ts`? Is there a way to decouple the two types? I'm not even sure why `lib.dom` is being included in this context.

**Comment by peterblazejewicz** [comment]
I assume this is because some setups use, say:
```ts
/// <reference types="node" />
```
(a global reference)
and within tsconfig:
```json
lib: [ "dom" ]
```
this results in both being used to resolve the same symbol.
Old problem without any clear solution as of yet. Simon, which done a ton of work for @types/node recently, created this TS issue:
https://github.com/microsoft/TypeScript/issues/43972

**Comment by yoursunny** [comment]
How about making `@types/web` a peer dependency of `@types/node`, then importing types from that library?
This ensures consistency between typings, for projects that target both Node and browsers.

**Review by mcollina — APPROVED** [review]
lgtm

**Comment by zone117x** [comment]
@peterblazejewicz 
> Old problem without any clear solution as of yet. Simon, which done a ton of work for @types/node recently, created this TS issue:
> [microsoft/TypeScript#43972](https://github.com/microsoft/TypeScript/issues/43972)

Ah, looks like the issue keeps getting pushed as well. I've been looking at how this is handled in similar scenarios, for example with the `URL` & `URLSearchParams` globals:
https://github.com/DefinitelyTyped/DefinitelyTyped/blob/ee088b625fed6feb25309a89964a9f27c65c06b1/types/node/url.d.ts#L860-L887

Does an approach like this make sense for `AbortController` & `AbortSignal`? One potential issue is that since the DOM definition doesn't include `AbortSignal.timeout(number)`, this workaround would make it so that the projects that reference the dom lib will use the dom `AbortSignal` definition rather than the `@types/node` one (if I am understanding this workaround).


@yoursunny
> How about making `@types/web` a peer dependency of `@types/node`, then importing types from that library?
> This ensures consistency between typings, for projects that target both Node and browsers.

Interesting, is there precedent for doing something like this, especially in this package?

**Comment by zone117x** [comment]
@antongolub it looks like you have experience with this situation (like in https://github.com/DefinitelyTyped/DefinitelyTyped/pull/58619) -- any suggestions here?

**Comment by yoursunny** [comment]
> > How about making `@types/web` a peer dependency of `@types/node`, then importing types from that library?
> > This ensures consistency between typings, for projects that target both Node and browsers.
> 
> Interesting, is there precedent for doing something like this, especially in this package?

No precedent, but this could be a solution.

A drawback is that the typing would not be able to capture the difference between web standards and Node implementation.
If a certain method or property is available in browsers but unavailable in Node, the typing would still indicate that it is available.
But then, this situation already exists when developer needs to do runtime feature testing to detect browser support for methods and properties that are not universally available.
They just have to do the same sort of runtime feature testing in Node instead of expecting the typing to accurately match the running Node version.

**Comment by zone117x** [comment]
@rbuckton do you have any recommendations for a workaround here?

**Comment by rbuckton** [comment]
Unfortunately, I think the only solution is to update `lib.dom` in TypeScript to also have `AbortSignalConstructor`. Starting in TS 4.5 it's now possible to override `lib.dom` via a package dependency, but I don't think that would help in this case.

**Pull request closed without merging**